### PR TITLE
CFSQL-1589 migrations_pattern for configuring recursive migration discovery 

### LINKED ISSUE #13257: D1 `migrations_dir` | Support recursive `.sql` file search
### What versions & operating system are you using?

v4.62 - MacOS

### Please provide a link to a minimal reproduction

https://github.com/drizzle-team/drizzle-orm/issues/5266

### Describe the Bug

As per [this issue](https://github.com/drizzle-team/drizzle-orm/issues/5266) for drizzle-orm, with their upcoming v1, the migration files live in individual directories now.

Currently, wrangler only searches for `.sql` files in the provided `migrations_dir` exactly without searching subdirectories, resulting in it not finding the migrations it should apply.

While there are a variety of approaches to resolve this, I feel the simplest would be to search all subdirectories as well. Alternatively, glob matching might be more powerful.

### Please provide any relevant error logs

_No response_

### PR DESCRIPTION
**Pull request by alsuren**
Fixes  #13257 (CFSQL-1589).

Adds a `migrations_pattern` config field to D1 bindings so users can
opt their D1 migrations into a nested layout (drizzle's
`migrations/*/migration.sql`, prisma-style folders, custom
extensions, …) instead of the implicit `migrations/*.sql`. Without
`migrations_pattern`, behaviour is unchanged.

Also includes a commit to fix our logging, because it was causing a bunch of test failures as I added more tests (best reviewed with whitespace hidden: https://github.com/cloudflare/workers-sdk/pull/14089/changes/54d293facacc02fc5037dcf2ec3ae6ebf36f8cee?w=1 .

---

- Tests
  - [x] Tests included/updated
  - [ ] Automated tests not possible - manual testing has been completed as follows:
  - [ ] Additional testing not necessary because:
- Public documentation
  - [x] Cloudflare docs PR(s): https://github.com/cloudflare/cloudflare-docs/pull/31140
  - [ ] Documentation not necessary because:

*A picture of a cute animal (not mandatory, but encouraged)*

```
 ,_     _
 |\\_,-~/
 / _  _ |    ,--.
(  @  @ )   / ,-'
 \  _T_/-._( (
 /         `. \
|         _  \ |
 \ \ ,  /      |
  || |-_\__   /
 ((_/`(____,-'
```
-- https://www.asciiart.eu/art/64248696ab474307

### CODE REVIEW

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- @cloudflare/wrangler
- ✅ @cloudflare/d1

<details><summary>Show detailed file reviewers</summary>

- .changeset/d1-execute-logger-level.md: [@cloudflare/wrangler]
- .changeset/d1-migrations-pattern.md: [@cloudflare/wrangler]
- packages/workers-utils/src/config/environment.ts: [@cloudflare/wrangler]
- packages/workers-utils/src/config/validation.ts: [@cloudflare/wrangler]
- packages/workers-utils/src/worker.ts: [@cloudflare/wrangler]
- packages/wrangler/src/__tests__/d1/migrate.test.ts: [@cloudflare/wrangler]
- packages/wrangler/src/__tests__/d1/migrations/helpers.test.ts: [@cloudflare/wrangler]
- packages/wrangler/src/__tests__/d1/utils.test.ts: [@cloudflare/wrangler]
- packages/wrangler/src/d1/execute.ts: [@cloudflare/wrangler]
- packages/wrangler/src/d1/migrations/apply.ts: [@cloudflare/wrangler]
- packages/wrangler/src/d1/migrations/create.ts: [@cloudflare/wrangler]
- packages/wrangler/src/d1/migrations/helpers.ts: [@cloudflare/wrangler]
- packages/wrangler/src/d1/migrations/list.ts: [@cloudflare/wrangler]
- packages/wrangler/src/d1/types.ts: [@cloudflare/wrangler]
- packages/wrangler/src/d1/utils.ts: [@cloudflare/wrangler]

</details>

File: packages/workers-utils/src/config/environment.ts [hunk 1]
```diff
@@ -990,6 +990,21 @@ export interface EnvironmentNonInheritable {
 		migrations_table?: string;
 		/** The path to the directory of migrations for this D1 database (defaults to './migrations'). */
 		migrations_dir?: string;
+		/**
+		 * A glob pattern (relative to the Wrangler config file) used to discover
+		 * migration files for this D1 database. Defaults to `${migrations_dir}/*.sql`
+		 * if not specified.
+		 *
+		 * Use this to opt in to nested layouts such as `migrations/*\/migration.sql`
+		 * (as produced by some ORMs).
```
Comment (FlorentCollin) [inline packages/workers-utils/src/config/environment.ts:999]: There is no need to escape the `/` in this case.
```suggestion
		 * Use this to opt in to nested layouts such as `migrations/*/migration.sql`
```

File: packages/workers-utils/src/config/environment.ts [hunk 2]
```diff
@@ -990,6 +990,21 @@ export interface EnvironmentNonInheritable {
...
+		 *
+		 * When `migrations_pattern` is set, `migrations_dir` must also be set, and
+		 * `migrations_pattern` must start with `${migrations_dir}/`. This keeps the
+		 * relationship between the two settings explicit and lets Wrangler record
+		 * each migration's name in the `d1_migrations` table as a path relative to
```
Comment (FlorentCollin) [inline packages/workers-utils/src/config/environment.ts]: The migration table can have a different name if the user specified the `migrations_table` properties in their `wrangler.jsonc`.

File: packages/wrangler/src/d1/migrations/helpers.ts [hunk 3]
```diff
@@ -1,15 +1,127 @@
...
+	const configFile = configFileName(configPath);
+	const projectPath = path.dirname(configPath);
+
+	const rawDir = databaseInfo?.migrationsDirRaw;
+	const rawPattern = databaseInfo?.migrationsPattern;
+
+	if (rawPattern !== undefined && rawDir === undefined) {
+		throw new UserError(
+			`You have set \`migrations_pattern: "${rawPattern}"\` in your ${configFile} file but no \`migrations_dir\` for this D1 binding.\n\n` +
+				`When \`migrations_pattern\` is set, \`migrations_dir\` must also be set, and \`migrations_pattern\` must start with \`\${migrations_dir}/\`. Add a \`migrations_dir\` entry to your ${configFile} file (for example, \`migrations_dir: "migrations"\`).`,
```
Comment (FlorentCollin) [inline packages/wrangler/src/d1/migrations/helpers.ts]:
```suggestion
				`When \`migrations_pattern\` is set, \`migrations_dir\` must also be set, and \`migrations_pattern\` must start with \`\${migrations_dir}/\`. Add a \`migrations_dir\` entry to your ${configFile} file (for example, \`"migrations_dir": "migrations"\`).`,
```

File: packages/wrangler/src/d1/migrations/helpers.ts [hunk 4]
```diff
@@ -1,15 +1,127 @@
...
+	const migrationsDir = normalizeRelativePath(
+		databaseInfo?.migrationsFolderPath ?? DEFAULT_MIGRATION_PATH
+	);
+
+	if (rawPattern !== undefined) {
+		const normalizedPattern = normalizeRelativePath(rawPattern);
+		if (!normalizedPattern.startsWith(`${migrationsDir}/`)) {
+			const suggestedPattern = getDefaultMigrationsPattern(migrationsDir);
+			throw new UserError(
+				`The configured \`migrations_pattern: "${rawPattern}"\` in your ${configFile} file must start with \`${migrationsDir}/\` to match \`migrations_dir: "${migrationsDir}"\`.\n\n` +
```
Comment (FlorentCollin) [inline packages/wrangler/src/d1/migrations/helpers.ts]:
```suggestion
				`The configured \`migrations_pattern: "${rawPattern}"\` in your ${configFile} file must start with \`${migrationsDir}/\` to match \`"migrations_dir": "${migrationsDir}"\`.\n\n` +
```

File: packages/wrangler/src/d1/migrations/helpers.ts [hunk 5]
```diff
@@ -1,15 +1,127 @@
...
+ * Normalize a relative path or glob into a canonical form for string-prefix
+ * comparisons:
+ *
+ *  - Backslashes flipped to forward slashes.
+ *  - Leading `./` and `//` runs collapsed (via `path.posix.normalize`).
+ *  - Trailing `/` stripped (`normalize("foo/")` keeps it; we don't want it).
+ */
+export function normalizeRelativePath(p: string): string {
+	const forwardSlashed = p.replace(/\\/g, "/");
+	const normalized = path.posix.normalize(forwardSlashed);
```
Comment (FlorentCollin) [inline packages/wrangler/src/d1/migrations/helpers.ts:145]: I hope that people won't start using symlinks because I think it's possible to break everything 😅.

File: packages/wrangler/src/d1/migrations/helpers.ts [hunk 6]
```diff
@@ -130,33 +243,182 @@ const listAppliedMigrations = async ({
...
+	const drizzleFiles = listFilesRelative(
+		walkRoot,
+		new Minimatch("*/migration.sql", { dot: false })
+	);
+	if (drizzleFiles.length === 0) {
+		return;
+	}
+	const drizzlePattern = migrationsDir
+		? `${migrationsDir}/*/migration.sql`
+		: "*/migration.sql";
```
Comment (FlorentCollin) [inline packages/wrangler/src/d1/migrations/helpers.ts]: Is it possible for `migrationsDir` to be falsy? The recommended pattern if it is falsy doesn't follow the rules stating that `migrations_pattern` must start by the value of the `migrations_dir`.

File: packages/wrangler/src/d1/migrations/helpers.ts [hunk 7]
```diff
@@ -130,33 +243,182 @@ const listAppliedMigrations = async ({
...
+ * into a subdirectory we ask whether its relative path could be a prefix of
+ * something matching `matcher.pattern`. If not, we skip the descent. So a
+ * `*.sql` pattern never recurses, `*\/migration.sql` only descends one
+ * level, `**\/*.sql` recurses unconditionally.
+ */
+function listFilesRelative(dir: string, matcher: Minimatch): string[] {
+	const out: string[] = [];
+	const stack: Array<{ abs: string; rel: string }> = [{ abs: dir, rel: "" }];
+
+	while (stack.length > 0) {
```
Comment (FlorentCollin) [inline packages/wrangler/src/d1/migrations/helpers.ts:292]: ~Would it be possible to replace the manual recursive file walker if you use `fs.readdirSync(abs, { withFileTypes: true, recursive: true })`?
https://nodejs.org/api/fs.html#fsreaddirsyncpath-options~

After reading the tests I now understand that it's because we don't want to recursively walk directories that don't are never going to match the pattern.

File: packages/wrangler/src/d1/migrations/helpers.ts
Comment (FlorentCollin) [inline packages/wrangler/src/d1/migrations/helpers.ts:1]: For the error/warning messages, should we also change the separator to use the platform separator (`\` in Windows and `/` in posix compliant platforms)?

File: packages/wrangler/src/__tests__/d1/migrations/helpers.test.ts [hunk 8]
```diff
@@ -41,35 +83,894 @@ describe("getMigrationNames", () => {
...
+			"migrations/README.md",
+			"migrations/test_data/destroy.sql",
+		]);
+		expect(getNextMigrationNumber(migrationsConfig())).toEqual(2);
+	});
+
+	it("collapses multiple files inside a single numbered directory to one number", ({
+		expect,
+	}) => {
+		// `0001_init/` is a single migration (the directory itself); even
```
Comment (FlorentCollin) [inline packages/wrangler/src/__tests__/d1/migrations/helpers.test.ts:563]: What happens if the user mix migrations directories like this:
```
- migrations/0001_init/migration.sql
- migrations/something/0002_cleanup.sql
```
Is the next migration number 2 or 3? Does it even matter?

**Comment by FlorentCollin** [comment]
It was not easy to review, but it's absolutely not your fault. Thanks for taking the time to write it - I really like the direction.

File: packages/wrangler/src/d1/migrations/helpers.ts (hunk 6 above)
Comment (lambrospetrou) [reply packages/wrangler/src/d1/migrations/helpers.ts]: Does the pattern work if it's `./*/migration.sql` instead of starting with `*`? Seems clearer and the wildcard prefix might be interpreted "every folder" not necessarily relative from here.

Also, if the migrations_dir is false, shouldn't we error earlier than here?

File: packages/workers-utils/src/config/environment.ts (hunk 1 above)
Comment (alsuren) [reply packages/workers-utils/src/config/environment.ts:999]: there is, because `*/` is how you finish a multi-line comment

File: packages/wrangler/src/__tests__/d1/migrations/helpers.test.ts (hunk 8 above)
Comment (alsuren) [reply packages/wrangler/src/__tests__/d1/migrations/helpers.test.ts:563]: 2

only top level numbers are considered, because d1 migrations create can only create top level migrations

File: packages/wrangler/src/d1/migrations/helpers.ts
Comment (alsuren) [reply packages/wrangler/src/d1/migrations/helpers.ts:1]: I feel like wrangler configs should use unix style paths everywhere for portability. Wrangler seems to be doing fine resolving unix style paths on windows, according to the test suite. This is what I'm used to from python-land as well.

File: packages/wrangler/src/d1/migrations/helpers.ts (hunk 5 above)
Comment (alsuren) [reply packages/wrangler/src/d1/migrations/helpers.ts:145]: I guess they get to keep the pieces in that case.

File: packages/wrangler/src/d1/migrations/helpers.ts (hunk 6 above)
Comment (alsuren) [reply packages/wrangler/src/d1/migrations/helpers.ts]: - [x] I will add "migrationsDir is nonempty" to my list of invariants when parsing and applying defaults, and kill off this ternary

Comment (alsuren) [reply packages/wrangler/src/d1/migrations/helpers.ts]: migrations_dir = "" is treated as the default `migrations` in main

migrations_dir = "." is allowed in main. I wrote a test to assert that this continues to work on this branch.

**Review by FlorentCollin — APPROVED** [review]

File: packages/wrangler/src/d1/migrations/helpers.ts [hunk 9]
```diff
@@ -1,15 +1,130 @@
...
+}): MigrationsConfig {
+	const configFile = configFileName(configPath);
+	const projectPath = path.dirname(configPath);
+
+	const rawDir = databaseInfo?.migrationsDirRaw;
+	const rawPattern = databaseInfo?.migrationsPattern;
+
+	if (rawPattern !== undefined && rawDir === undefined) {
+		throw new UserError(
+			`You have set \`migrations_pattern: "${rawPattern}"\` in your ${configFile} file but no \`migrations_dir\` for this D1 binding.\n\n` +
```
Comment (james-elicx) [inline packages/wrangler/src/d1/migrations/helpers.ts]:
```suggestion
			`You have set \`migrations_pattern: "${rawPattern}"\` in your ${configFile} file but have not set \`migrations_dir\` for this D1 binding.\n\n` +
```

File: packages/wrangler/src/d1/execute.ts
Comment (james-elicx) [inline packages/wrangler/src/d1/execute.ts:132]: does this bit need the log level restoration as well, as it's outside the try catch?

File: packages/wrangler/src/d1/migrations/helpers.ts [hunk 10]
```diff
@@ -1,15 +1,130 @@
...
+		);
+	}
+
+	const migrationsDir = normalizeRelativePath(
+		databaseInfo?.migrationsFolderPath ?? DEFAULT_MIGRATION_PATH
+	);
+
+	if (rawPattern !== undefined) {
+		const normalizedPattern = normalizeRelativePath(rawPattern);
+		if (!normalizedPattern.startsWith(`${migrationsDir}/`)) {
```
Comment (james-elicx) [inline packages/wrangler/src/d1/migrations/helpers.ts]: if someone sets `.` or `./` as their directory and pattern, i think this would throw wouldn't it?

`.` and `./` -> `.` after normalization. `.` would never start with `./`, and therefore this check would fail and an error would throw, even though technically the values are the same, unless i'm missing something.

File: packages/wrangler/src/d1/migrations/helpers.ts
Comment (james-elicx) [inline packages/wrangler/src/d1/migrations/helpers.ts:173]: i was reading through some more of this file - if the migrations dir in the config is set to `./migrations`, that would be normalised to `migrations` and match here wouldn't it? if so, i believe this warning would be misleading

File: packages/wrangler/src/d1/migrations/helpers.ts [hunk 11]
```diff
@@ -130,33 +246,181 @@ const listAppliedMigrations = async ({
...
+ * segment, falling back to lex order on ties. Numbered files sort before
+ * unnumbered ones.
+ *
+ * Numeric ordering matters for users with inconsistently-padded numeric
+ * prefixes (`1_a.sql`, `9_b.sql`, `10_c.sql`); a pure lex sort would put
+ * `10_c.sql` between `1_a.sql` and `9_b.sql`.
+ */
+export function compareMigrationPaths(a: string, b: string): number {
+	const aNum = leadingMigrationNumber(a);
+	const bNum = leadingMigrationNumber(b);
```
Comment (james-elicx) [inline packages/wrangler/src/d1/migrations/helpers.ts:342]: i'm probably getting a bit too nitpicky now, but i was wondering, if someone has nested subdirectories which contain numerically-ordered migration files, then in theory, the directories would be ordered correctly, but the files inside those directories might not be ordered numberically, since this is only based on the first segment? is that understanding correct? maybe we need to go into each segment during sorting and compare at each level?

e.g. `./migrations/0001_posts/1_a.sql` and `./migrations/0001_posts/9_b.sql` and `./migrations/0001_posts/10_c.sql` with dir `./migrations` and pattern `./migrations/**/*.sql`

or maybe this isn't realistic?

Comment (alsuren) [reply packages/wrangler/src/d1/migrations/helpers.ts:342]: Yeah, I guess that would be reasonable and consistent. 

We don't currently allow this, so whatever we pick here will be the behaviour that people depend on. I will change it to do that instead.

File: packages/wrangler/src/d1/migrations/helpers.ts (hunk 10 above)
Comment (alsuren) [reply packages/wrangler/src/d1/migrations/helpers.ts]: yeah, good catch. I initially only considered . as an afterthought after realising that it was supported in main. Should be fixed now.

**Review by james-elicx — APPROVED** [review]

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- ✅ @cloudflare/d1
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>


</details>

**Review by workers-devprod — APPROVED** [review]
Codeowners reviews satisfied
